stdlib/Random/src/RNGs.jl

```diff
@@ -285,14 +285,57 @@ function seed!(r::MersenneTwister, seed::Vector{UInt32})
     return r
 end
 
-seed!(r::MersenneTwister=GLOBAL_RNG) = seed!(r, make_seed())
+seed!(r::MersenneTwister=get_local_rng()) = seed!(r, make_seed())
 seed!(r::MersenneTwister, n::Integer) = seed!(r, make_seed(n))
-seed!(seed::Union{Integer,Vector{UInt32}}) = seed!(GLOBAL_RNG, seed)
+seed!(seed::Union{Integer,Vector{UInt32}}) = seed!(get_local_rng(), seed)
```

Comment:
One would naively expect `Random.seed!(i::Int)` to seed all the global RNGs, not just the one of the calling thread.

Reply:
This leads to the question how we avoid correlated rngs. If we seed them from the same point then there is a reasonable likelihood that a user writes:
```
Base.seed!(1)
@threads for i in 1:nthreads()
    @show rand()
end
```
and get the same same stream of rngs on each thread.

Reply:
I think that's what chethega was suggesting is the desired behavior

Reply:
No is saying that it should seed all rng's not, that we should create correlated rngs.
I agree that `seed` should seed all rngs, but they should be initialized through `randjump` so that the seed provides reproducibility,
while giving you something useful if you want to do something like monte carlo.

Reply:
Very naive way: `seed!(s)` seeds all thread's MTs, and thread `k` is seeded with `mix(k, s)` (where `mix` can range from `xor` to full random function like e.g. aes). Speed is immaterial, since `seed!` is expected to be a slow operation.

Reply:
I'm slightly sceptical that it's a good idea to call `randjump` implicitly here. Also, `randjump` currently (that is, for `MersenneTwister`) costs about 20ms on my machine, with 16 cores it will costs 1/3 s, when maybe the user doesn't need at all to initialize all the global RNGs.

Reply:
Can/will `nthreads()` (be able to) be changed after Julia has started? (In this case it may be error prone to initialize all RNGs, as after a change of `nthreads()`, the new RNGs wouldn't be initialized).

Reply:
Currently no, but if we add that feature we would have to address this.

Reply:
>Also, randjump currently (that is, for MersenneTwister) costs about 20ms

Then we should just not call randjump, and instead call something like `seed!(mt::MersenneTwister, xor(user_seed, thread_to_reseed<<8))`. This one is a handful of us, not ms.

Reply:
Perhaps we should use a SHA hash of a tuple of seeds and use `(task, seed)` as the implicit seed in each task. That gives use very uncorrelated streams without relying on jumping. I think we should be doing this kind of secure hashing for seeding anyway since otherwise there are major problems with using consecutive seeds with small numbers of bits set, e.g. `seed!(1)`.

Reply:
You mean `(threadid(), seed)`, not task, right?

Otherwise I agree. Note that this changes all random streams, though.

Reply:
Re-open: This behavior needs warnings in the docs. A naive user could otherwise expect `seed!(seed::Int) = foreach(1:nthreads()) do tid begin seed!(get_local_rng(tid), seed)) end end`. 

That is, naive use would call `seed!` in the initial single-threaded part of a script, and expect this to help reproducibility, regardless of whether later parts of the script call into multi-threaded libraries.

Reply:
> expect this to help reproducibility

Your naive use or suggestion won't help do this for code that uses threads. With threads, there's no concept of reproducibility or even Turing-equivalence anymore, since there's no linearization of the program. I don't think there's anything to document. In fact, it would be even worse than this if we did implement your suggestion, since then a naive usage of `seed!` in a MT program could ruin an otherwise predictable run of concurrent `rand` calls on another thread, and would be a gotcha for future applications.

Reply:
I think it makes sense to put a note in the docs about this, since it's not obvious what `seed!()` means in the presence of threads. Somebody might reasonably guess that it's thread-local, but it's helpful to spell it out.